### PR TITLE
Functional test

### PR DESCRIPTION
**Pull request by LuitelSamikshya**

### CODE REVIEW

File: v2/cmd/functional-test/testcases.txt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+example.com {{binary}} -silent
+example.com {{binary}} -silent -il
+example.com {{binary}} -silent l test-data/request.txt
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: I think it should be `-l`

File: .github/workflows/functional-test.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  functional:
+    name: Functional Test
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17
+
+      - name: Install libpcap-dev
```
Comment (Mzack9999) [inline .github/workflows/functional-test.yml:18]: We need another step to install `nmap`:

```
- name: Install nmap
        run: sudo apt install nmap
```

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
The box setup should include installing and starting in the background one or more generic listeners like `https://github.com/projectdiscovery/simplehttpserver` and the scan targets should be limited only to `127.0.0.1` and those specific ports

File: .github/workflows/functional-test.yml (hunk 2 above)
Comment (LuitelSamikshya) [reply .github/workflows/functional-test.yml:18]: Done.

File: v2/cmd/functional-test/testcases.txt (hunk 1 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done.

File: v2/cmd/functional-test/testcases.txt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+127.0.0.1 {{binary}} -silent -p 8000
+127.0.0.1 {{binary}} -silent -il
+127.0.0.1 {{binary}} -silent -l test-data/request.txt -p 8000
+127.0.0.1 {{binary}} -silent -p 8000,443
+127.0.0.1 {{binary}} -silent -tp 100 -p 8000
+127.0.0.1 {{binary}} -silent -ep 200  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -ec 100,400  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -version
+127.0.0.1 {{binary}} -silent -retries 5  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -c 25  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -verify  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -timeout 30  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -r test-data/request.txt  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -warm-up-time 2  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

File: v2/cmd/functional-test/testcases.txt [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+127.0.0.1 {{binary}} -silent -nmap-cli 'nmap -sV'  -p 8000
```
Comment (Mzack9999) [inline v2/cmd/functional-test/testcases.txt]: Please remove the double space between CLI args

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
Looks good, a couple of notes:
- [x] Please remove where pinpointed all extra spaces between cli args
- [x] In each test case it was specified `-silent` buy the verbosity used should be the one defined in `run.sh`. With `Debug=true ./run.sh` the binary would be invoked with both `-debug` and `-silent` flags

File: v2/cmd/functional-test/testcases.txt (hunk 3 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 4 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 5 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 6 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 7 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 8 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 10 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 9 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 11 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 3 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: done

File: v2/cmd/functional-test/testcases.txt (hunk 4 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 5 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 6 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 7 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 8 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 9 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 10 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

File: v2/cmd/functional-test/testcases.txt (hunk 11 above)
Comment (LuitelSamikshya) [reply v2/cmd/functional-test/testcases.txt]: Done

**Review by Mzack9999 — APPROVED** [review]

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**